### PR TITLE
Add integration for Veeam Backup & Replication REST API

### PR DESCRIPTION
**Pull request by nikita-konoplin**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A few sentences describing the overall goals of the pull request's commits.

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by samuelFain** [comment]
@ nikita-konoplin Thank you for your contribution!
The code review is on-going and we are doing our best to post it asap, we a your penitence!

File: Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI_image.png
Comment (samuelFain) [inline Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI_image.png:1]: Please add relevant integration image.
See [here](https://xsoar.pan.dev/docs/integrations/integration-logo) for image requirements.

**Review by samuelFain — CHANGES_REQUESTED** [review]
@nikita-konoplin Very nice!
The PR looks great! Please see my few comments.

Also, please merge your base branch form content/master, and then pull from your base branch to this branch. This is needed due to a change made in environment configuration files of content to allow all the Github actions checks to run in this PR.

Thanks!

File: Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI_image.png
Comment (nikita-konoplin) [reply Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI_image.png:1]: Logo images are not ready yet, we will attach them when ready

File: Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI.py
Comment (nikita-konoplin) [reply Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI.py]: done

File: Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI.py
Comment (nikita-konoplin) [reply Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI.py]: done

File: Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI.py
Comment (nikita-konoplin) [reply Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI.py]: done

**Comment by nikita-konoplin** [comment]
@samuelFain Thank you!
Logo images are not ready yet, we will attach them when ready.

**Review by samuelFain — CHANGES_REQUESTED** [review]
@nikita-konoplin Great job!
Few more points from our automated checks:
1. `demisto-sdk secrets` had identified suspicious values in the code (see results [here](https://app.circleci.com/pipelines/github/demisto/content/376435/workflows/745e3ee7-b5f0-4ec7-89d6-e0591af38481/jobs/750929/parallel-runs/0/steps/0-118)), please examine them. In case these values are not secrets, please ignore them as described in https://xsoar.pan.dev/docs/concepts/demisto-sdk#secrets.
2. Could not find pack-ignore file at path /home/circleci/project/Packs/Veeam/.pack-ignore for pack Veeam. Please add pack-ignore file.
3. `pack_metadata.json` not found in pack in 'Packs/Veeam'. I think the metadata file exist in https://github.com/demisto/content/pull/35706, it should be moved to this PR, as it defines the entire file structure of the pack for both PRs.

For more information see `ci/circleci: Run Validations` and `pre-commit (pull_request)` Github actions results.

File: Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI_image.png
Comment (nikita-konoplin) [reply Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI_image.png:1]: integration image was added

**Comment by nikita-konoplin** [comment]
@samuelFain Thank you!
I have added all the files needed.

**Comment by samuelFain** [comment]
@nikita-konoplin Amazing work so far!
There are more issues from our automated PR checks post your last changes, please see details below:

- Unit-tests for 'Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI.py' must reach a coverage of at least 70.0 percent (currently: 47.36 percent). You can use the following guide - https://xsoar.pan.dev/docs/integrations/unit-testing as well as examples of unit tests in the HelloWorld integration and other packs from the content repository.
- The integration is a fetch integration and is missing the following mandatory params: `max_fetch`.
- Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI.yml: [DO106] - docker image demisto/python3:3.10.14.92207's tag 3.10.14.92207 is outdated. The latest tag is 3.11.9.107902
- Packs/Veeam/README.md: [PA100] - "README.md" file does not exist, create one in the root of the pack
- All commands in the .yml file are missing the "description" field.
-  Content item 'VBR REST API' using content items: 'Veeam Backup & Replication incoming mapper', 'Veeam Backup & Replication Incidents Classifier' which cannot be found in the repository.
- Veeam/Author_image.png: [IM109] - Partners must provide a non-empty author image under the path /home/circleci/project/Packs/Veeam/Author_image.png.
See example:
image


You can check `ci/circleci: Run Validations` and `pre-commit (pull_request)` Github actions results for more information.

**Comment by nikita-konoplin** [comment]
Hi @samuelFain, @edik24  
Thank you for review!

Maybe I will add VBR integration into the second PR (https://github.com/demisto/content/pull/35706) so, we will have all pack files, mappers, readme, etc.? We will have one PR for our pack, and we will not need to find and copy files every time.
What do you think?

**Comment by samuelFain** [comment]
> Hi @samuelFain, @edik24 Thank you for review!
> 
> Maybe I will add VBR integration into the second PR (#35706) so, we will have all pack files, mappers, readme, etc.? We will have one PR for our pack, and we will not need to find and copy files every time. What do you think?

@nikita-konoplin I don't think its necessary, just move the following files from https://github.com/demisto/content/pull/35706 to this PR so the Pack and integration structure is complete:
- Packs/Veeam/Author_image.png
- Packs/Veeam/README.md
-  'Veeam Backup & Replication incoming mapper', 'Veeam Backup & Replication Incidents Classifier' 

Basically, in this PR we are creating the necessary Pack structure, along with the `Veeam Backup & Replication REST API` integration inside it, once it is merged, we can merge https://github.com/demisto/content/pull/35706 as well into the already existing Pack.

**Comment by nikita-konoplin** [comment]
> @nikita-konoplin Amazing work so far! There are more issues from our automated PR checks post your last changes, please see details below:
> 
> * Unit-tests for 'Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI.py' must reach a coverage of at least 70.0 percent (currently: 47.36 percent). You can use the following guide - https://xsoar.pan.dev/docs/integrations/unit-testing as well as examples of unit tests in the HelloWorld integration and other packs from the content repository.
> * The integration is a fetch integration and is missing the following mandatory params: `max_fetch`.
> * Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI.yml: [DO106] - docker image demisto/python3:3.10.14.92207's tag 3.10.14.92207 is outdated. The latest tag is 3.11.9.107902
> * Packs/Veeam/README.md: [PA100] - "README.md" file does not exist, create one in the root of the pack (Please move it from [Add remaining parts of the Veeam content pack #35706](https://github.com/demisto/content/pull/35706) to this PR).
> * Please add missing the "description" field for commands in the .yml file.
> * Content item 'VBR REST API' using content items: 'Veeam Backup & Replication incoming mapper', 'Veeam Backup & Replication Incidents Classifier' which cannot be found in the repository.
> * Veeam/Author_image.png: [IM109] - Partners must provide a non-empty author image under the path /home/circleci/project/Packs/Veeam/Author_image.png. (Please move it from [Add remaining parts of the Veeam content pack #35706](https://github.com/demisto/content/pull/35706) to this PR).
> 
> You can check `ci/circleci: Run Validations` and `pre-commit (pull_request)` Github actions results for more information.

Hi @samuelFain, Thank you for review!

* The integration is a fetch integration and is missing the following mandatory params: max_fetch.

- In our fetch function, we call different endpoints and create three types of incidents. There can be only one incident "Configuration Backup", and other two have its max limit parameters: "Backup Repository Events Per Request", "Malware Events Per Request". We do not know how to apply the `max_fetch` parameter here, because if "Backup Repository Events Request" will return us N >=`max_fetch` events, "Malware Events" will not be created at all. Can we skip this check?

**Comment by nikita-konoplin** [comment]
> > Hi @samuelFain, @edik24 Thank you for review!
> > Maybe I will add VBR integration into the second PR (#35706) so, we will have all pack files, mappers, readme, etc.? We will have one PR for our pack, and we will not need to find and copy files every time. What do you think?
> 
> @nikita-konoplin I don't think its necessary, just move the following "Pack level" files from #35706 to this PR so the Pack and integration structure is complete:
> 
> * Packs/Veeam/Author_image.png
> * Packs/Veeam/README.md
> * Packs/Veeam/Classifiers
> 
> Basically, in this PR we are creating the necessary Pack structure, along with the `Veeam Backup & Replication REST API` integration inside it, once it is merged, we can merge #35706 as well into the already existing Pack.

Hi @samuelFain,
Required files were added.

**Comment by samuelFain** [comment]
> > @nikita-konoplin Amazing work so far! There are more issues from our automated PR checks post your last changes, please see details below:
> > 
> > * Unit-tests for 'Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI.py' must reach a coverage of at least 70.0 percent (currently: 47.36 percent). You can use the following guide - https://xsoar.pan.dev/docs/integrations/unit-testing as well as examples of unit tests in the HelloWorld integration and other packs from the content repository.
> > * The integration is a fetch integration and is missing the following mandatory params: `max_fetch`.
> > * Packs/Veeam/Integrations/VBRRESTAPI/VBRRESTAPI.yml: [DO106] - docker image demisto/python3:3.10.14.92207's tag 3.10.14.92207 is outdated. The latest tag is 3.11.9.107902
> > * Packs/Veeam/README.md: [PA100] - "README.md" file does not exist, create one in the root of the pack (Please move it from [Add remaining parts of the Veeam content pack #35706](https://github.com/demisto/content/pull/35706) to this PR).
> > * Please add missing the "description" field for commands in the .yml file.
> > * Content item 'VBR REST API' using content items: 'Veeam Backup & Replication incoming mapper', 'Veeam Backup & Replication Incidents Classifier' which cannot be found in the repository.
> > * Veeam/Author_image.png: [IM109] - Partners must provide a non-empty author image under the path /home/circleci/project/Packs/Veeam/Author_image.png. (Please move it from [Add remaining parts of the Veeam content pack #35706](https://github.com/demisto/content/pull/35706) to this PR).
> > 
> > You can check `ci/circleci: Run Validations` and `pre-commit (pull_request)` Github actions results for more information.
> 
> Hi @samuelFain, Thank you for review!
> 
> * The integration is a fetch integration and is missing the following mandatory params: max_fetch.
> 
> * In our fetch function, we call different endpoints and create three types of incidents. There can be only one incident "Configuration Backup", and other two have its max limit parameters: "Backup Repository Events Per Request", "Malware Events Per Request". We do not know how to apply the `max_fetch` parameter here, because if "Backup Repository Events Request" will return us N >=`max_fetch` events, "Malware Events" will not be created at all. Can we skip this check?
> 
> image

@nikita-konoplin I see what you mean, but this check is a blocker in terms of merging this PR, as this is a content-wide convention.
My solution would be to add this parameter with the "hidden" value, making this parameter invisible in the instance configuration window.

Also, I'm still seeing errors from automated checks in the lines of:

> [ERROR] /home/circleci/project/Packs/Veeam/Classifiers/classifier-VeeamONERESTAPI.json: [GR103] - Content item 'Veeam One Incidents Classifier' using content items: 'Immutability Change Tracking', 'Job Duration', 'Enterprise Application Backup', 'Malware Detection Change Tracking', 'Immutability State', 'Job Duration Deviation (Veeam Backup for Microsoft 365)', 'Virtual Machine Backup', 'Unclassified', 'Backup Copy Creation Time', 'Physical Machine Backup', 'Possible Malware Activity', 'Backup Server Security Status', 'Job Disabling', 'Incremental Backup Size', 'Virtual Machine Replica', 'Incident Fetch Error' which cannot be found in the repository.

I think we dont have a choice but to combine both PRs in order to fix these issues, as you suggested earlier.
@edik24 FYI.

**Comment by nikita-konoplin** [comment]
Hi @samuelFain, @edik24  Thank you for review!

We've added the remaining parts of the Veeam content pack.
We can't run pre-commit locally (it just hangs), so we can't see test coverage for our integrations, but more tests have been added so it should be fine.

File: Packs/Veeam/pack_metadata.json
Comment (samuelFain) [inline Packs/Veeam/pack_metadata.json:2]: Please change to:
```suggestion
    "name": "Veeam",
```
This should fix the reoccurring IF113 validation error for all the json files.

**Review by samuelFain — CHANGES_REQUESTED** [review]
@nikita-konoplin Everything looks much better after combining the two PRs!

About the UT coverage, last check confirms your coverage is sufficient at 79.01%. So you are good on that part.

The **only** error I'm still getting is from validation IF113, which states:

> [2024-08-20 17:51:20,699] [ERROR] Packs/Veeam/IncidentFields/incidentfield-Veeam_Number_of_Stored_Restore_Points.json: [IF113] - Field name: Veeam Number of Stored Restore Points is invalid. Field name must start with the relevant pack name.
[2024-08-20 17:51:20,700] [INFO] To fix the problem, add pack name prefix to the field name. You can use the pack name or one of the prefixes found in the itemPrefix field in the pack_metadata. Example: Veeam App for Palo Alto Networks XSOAR Veeam Number of Stored Restore Points.
Also, make sure to update the field id and cliName accordingly. Example: cliName: veeamappforpaloaltonetworksxsoarveeamnumberofstoredrestorepoints.

This is probably due to a naming mismatch between the pack's json files which expects the pack name to be "Veeam" but it is currently "Veeam App for Palo Alto Networks XSOAR" in `Packs/Veeam/pack_metadata.json`.
Please change the "name" key value in `Packs/Veeam/pack_metadata.json` to: `"name": "Veeam"` as I stated in my latest review comment.
I believe this should solve this issue.

Thanks!

File: Packs/Veeam/pack_metadata.json
Comment (nikita-konoplin) [reply Packs/Veeam/pack_metadata.json:2]: Hi @samuelFain, there is a package name approved by our management. We cannot change it.
Also, we don't want field names like this veeamappforpaloaltonetworksxsoarveeamnumberofstoredrestorepoints.
Can we ignore this validation in .pack-ignore?

Comment (samuelFain) [reply Packs/Veeam/pack_metadata.json:2]: @nikita-konoplin Fine by me.

Comment (nikita-konoplin) [reply Packs/Veeam/pack_metadata.json:2]: done

**Comment by nikita-konoplin** [comment]
Hi @samuelFain, 
Thank you for review!

The necessary changes have been made.

**Review by samuelFain — CHANGES_REQUESTED** [review]
@nikita-konoplin 
Last issues (not seeing anything else at the moment in the automated PR checks):

- There is another validation failure [RM100], its ignorable so you can either ignore or add a more detailed description to the relevant README files, your choice.
Relevant files:
Packs/Veeam/Scripts/getfoldername/README.md - [RM100]
Packs/Veeam/Scripts/gethostname/README.md - [RM100]
> [2024-08-21 14:13:28,357] [ERROR] Packs/Veeam/Scripts/getfoldername/README.md: [RM100] - Failed verifying README.md Error Message is: Your Pack README is too small (30 chars). Please move its content to the pack description or add more useful information to the Pack README. Pack README files are expected to include a few sentences about the pack and/or images.
File "/home/circleci/project/Packs/Veeam/Scripts/getfoldername/README.md", line 0

- Getting an error regarding the use of a content item called "Unclassified", which probably comes from `"defaultIncidentType": "Unclassified"`.
Relevant files:
Packs/Veeam/Classifiers/classifier-VeeamONERESTAPI.json
Packs/Veeam/Classifiers/classifier-Veeam_Backup_&_Replication_RESTAPI.json

> [2024-08-21 14:14:34,921] [ERROR] /home/circleci/project/Packs/Veeam/Classifiers/classifier-Veeam_Backup_&_Replication_RESTAPI.json: [GR103] - Content item 'Veeam Backup & Replication Incidents Classifier' using content items: 'Unclassified' which cannot be found in the repository.
[2024-08-21 14:14:34,922] [ERROR] /home/circleci/project/Packs/Veeam/Classifiers/classifier-VeeamONERESTAPI.json: [GR103] - Content item 'Veeam One Incidents Classifier' using content items: 'Unclassified' which cannot be found in the repository.

**Comment by nikita-konoplin** [comment]
> @nikita-konoplin Last issues (not seeing anything else at the moment in the automated PR checks):
> 
> * There is another validation failure [RM100], its ignorable so you can either ignore or add a more detailed description to the relevant README files, your choice.
>   Relevant files:
>   Packs/Veeam/Scripts/getfoldername/README.md - [RM100]
>   Packs/Veeam/Scripts/gethostname/README.md - [RM100]
> 
> > [2024-08-21 14:13:28,357] [ERROR] Packs/Veeam/Scripts/getfoldername/README.md: [RM100] - Failed verifying README.md Error Message is: Your Pack README is too small (30 chars). Please move its content to the pack description or add more useful information to the Pack README. Pack README files are expected to include a few sentences about the pack and/or images.
> > File "/home/circleci/project/Packs/Veeam/Scripts/getfoldername/README.md", line 0
> 
> * Getting an error regarding the use of a content item called "Unclassified", which probably comes from `"defaultIncidentType": "Unclassified"`.
>   Relevant files:
>   Packs/Veeam/Classifiers/classifier-VeeamONERESTAPI.json
>   Packs/Veeam/Classifiers/classifier-Veeam_Backup_&_Replication_RESTAPI.json
> 
> > [2024-08-21 14:14:34,921] [ERROR] /home/circleci/project/Packs/Veeam/Classifiers/classifier-Veeam_Backup_&_Replication_RESTAPI.json: [GR103] - Content item 'Veeam Backup & Replication Incidents Classifier' using content items: 'Unclassified' which cannot be found in the repository.
> > [2024-08-21 14:14:34,922] [ERROR] /home/circleci/project/Packs/Veeam/Classifiers/classifier-VeeamONERESTAPI.json: [GR103] - Content item 'Veeam One Incidents Classifier' using content items: 'Unclassified' which cannot be found in the repository.

Hi @samuelFain,
If we don't specify the default incident type as "Unclassified", the incident type will be chosen randomly. Also, in content packs 'Panorama', 'OpsGenie' is done exactly the same way.
"defaultIncidentType": "Unclassified"
Can we also ignore this validation?

In pre-commit validation I also see error:
Error: 024-08-22 08:57:00,939] [ERROR] Packs/Veeam: [PA101] - The currentVersion in the metadata (1.0.60) doesn't match the latest rn version (none).
We want to use our internal build version. Is it possible?

**Comment by samuelFain** [comment]
@nikita-konoplin, Great Demo!!
As for your latest comment in the PR:

> If we don't specify the default incident type as "Unclassified", the incident type will be chosen randomly. Also, in content packs 'Panorama', 'OpsGenie' is done exactly the same way.
"defaultIncidentType": "Unclassified"

Can we also ignore this validation?
Yes, it is ignorable - you can ignore it if you want.

___

>In pre-commit validation I also see error:
Error: 024-08-22 08:57:00,939] [ERROR] Packs/Veeam: [PA101] - The currentVersion in the metadata (1.0.60) doesn't match the latest rn version (none).
We want to use our internal build version. Is it possible?

Since you add a brand new pack, it should officially be version 1.0.0 (no release notes for new packs, only future PRs from this point should add release notes). See https://xsoar.pan.dev/docs/documentation/release-notes#i-added-a-new-pack-do-i-need-release-notes for more info. This is a system wide feature that needs to be inforced and i'm afraid you can't create a new pack with a different version .

___

>Should we also do something with this one from 'Partner-Approved label verification'?
ERROR: Label Partner-Approved was not added to PR: 35670. Please ask the partner to review the code and add the label after his approval.

Its a label to make sure that in the case of a community contribution to a partner's pack, they have approved the PR.
In this case, the pack is yours, so no need for extra validation. I'll add the label so that the check passes.

**Comment by nikita-konoplin** [comment]
@samuelFain Thanks for the review!
We fixed all validations and suggestions after the demo.
But please do not merge this pull request until we get final approval from our management.

**Review by samuelFain — APPROVED** [review]
@nikita-konoplin, the validation error is coming from an older version of the validation checker. I ran the new validation checker and everything passed.
I'll proceed to merge this PR.

Thank you for your patience and great job!!

**Pull request merged**
